src/nucleus/tt.ml

```diff
@@ -151,9 +151,23 @@ let assumptions_ty (Ty t) = assumptions_term t
 
 (** Generic fold on a term. The functions [atom], [bound] and
     [hyps] tell it what to do with atoms, bound variables, and
-    assumptions, respectively. *)
-let rec at_var atom bound hyps ~lvl {term=e';assumptions;loc} =
-  let assumptions = hyps ~lvl assumptions in
+    assumptions, respectively.
+
+    If the assumptions on a subterm do not change, this means the entire subterm does not change.
+
+    If [substitute] used [at_var] directly, this shortcut would be incorrect:
+    [x where x = x * x] with [x] some atom does not change the assumptions.
+
+    However we only use [at_var] in the following cases:
+    - instantiation, which changes bound variables to atoms
+    - abstraction, which changes atoms to bound variables
+    So if a variable which will change appears, it will be in the assumptions and the assumptions will change.
+    *)
```

Comment:
This seems excdeedingly dangerous to me. One day someone will use `at_var` in an unsafe way. Since we use `at_var` only in two cases, I would actually be in favor of eliminating `at_var` and doubling the code here. Or have A REALLY BIG WARNING explaining the precondition.